ws-server/src/server.rs

```diff
@@ -428,257 +473,83 @@ async fn background_task(
-									middleware.on_result(name, false, request_start);
-									middleware.on_response(request_start);
-								}
-							},
-							MethodKind::Async(callback) => match method.claim(name, &resources) {
-								Ok(guard) => {
-									let sink = sink.clone();
-									let id = id.into_owned();
-									let params = params.into_owned();
-
-									let fut = async move {
-										let result = (callback)(id, params, sink, conn_id, Some(guard)).await;
-										middleware.on_result(name, result, request_start);
-										middleware.on_response(request_start);
-									};
-
-									method_executors.add(fut.in_current_span().boxed());
-								}
-								Err(err) => {
-									tracing::error!(
-										"[Methods::execute_with_resources] failed to lock resources: {:?}",
-										err
-									);
-									sink.send_error(req.id, ErrorCode::ServerIsBusy.into());
-									middleware.on_result(name, false, request_start);
-									middleware.on_response(request_start);
-								}
-							},
-							MethodKind::Subscription(callback) => match method.claim(&req.method, &resources) {
-								Ok(guard) => {
-									let result = if let Some(cn) = bounded_subscriptions.acquire() {
-										let conn_state =
-											ConnState { conn_id, close_notify: cn, id_provider: &*id_provider };
-										callback(id, params, sink.clone(), conn_state, Some(guard))
-									} else {
-										sink.send_error(
-											req.id,
-											reject_too_many_subscriptions(bounded_subscriptions.max()),
-										);
-										false
-									};
-									middleware.on_result(name, result, request_start);
-									middleware.on_response(request_start);
-								}
-								Err(err) => {
-									tracing::error!(
-										"[Methods::execute_with_resources] failed to lock resources: {:?}",
-										err
-									);
-									sink.send_error(req.id, ErrorCode::ServerIsBusy.into());
-									middleware.on_result(name, false, request_start);
-									middleware.on_response(request_start);
-								}
-							},
-							MethodKind::Unsubscription(callback) => {
-								// Don't adhere to any resource or subscription limits; always let unsubscribing happen!
-								let result = callback(id, params, &sink, conn_id);
-								middleware.on_result(name, result, request_start);
-								middleware.on_response(request_start);
-							}
-						},
+				let fut = async move {
+					let call = CallData {
+						conn_id,
+						resources: &resources,
+						max_response_body_size,
+						max_log_length,
+						methods: &methods,
+						bounded_subscriptions,
+						sink: &sink,
+						id_provider: &*id_provider,
+						middleware,
+						request_start,
+					};
+
+					let (response, maybe_pending_sub_tx) = process_single_request(data, call).await;
+					middleware.on_response(&response.result, request_start);
+
+					let _ = sink.send_raw(response.result);
+
+					if let Some(pending_sub_tx) = maybe_pending_sub_tx {
+						pending_sub_tx.accept();
```

Comment:
>  1. don't like this pending_sub_tx thing here; it feels leaky (and is an odd one-off for just subscriptions). If we removed it without doing anything else, what would the consequences be? That the middleware is potentially called after another message has been sent down the wire? Do we care enough about that?

that's fair, the issue is that the message would never hit the middleware stuff then. It would just be answered so actually changed it a little bit to get rid of this mess and actually cloned the message to send back to both the `ws task` and `middleware´. As we kind of assume that subscription response will be short just a `subscription ID` that should be okay to clone.

> 2. It looks like we already know what the response for a subscription is going to be; it's always MethodResponse::response(id, &uniq_sub.sub_id, sink.max_response_size() as usize). Thus, do we actually need to try and get that result back from the accept() call to here? We just need to know when accept() is called perhaps, and then can build the identical response to give to middleware when needed? (not sure if that actually is useful to ponder or not!)

Yeah, just I did that but it means that we need to clone every subscription response but that should be quite small anyway so should be ok.

Reply:
Just had a quick look over; nice! It def feels cleaner to me now!